cli/src/search.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright 2023 Polyphene.
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::{
+    ffi::OsStr,
+    fs::{self, File, Metadata},
+    io::Read,
+    path::Path,
+};
+
+use anyhow::Context;
+use kythera_lib::{Abi, WasmActor};
+use walkdir::WalkDir;
+
+/// A test structure composed by the target Actor and its multiple tests.
+pub struct Test {
+    pub actor: WasmActor,
+    pub tests: Vec<WasmActor>,
+}
+
+/// Read a WebAssembly Actor from a given file path.
+fn read_actor<P: AsRef<Path>>(path: P) -> anyhow::Result<WasmActor> {
+    let path = path.as_ref();
+    let mut file =
+        File::open(path).with_context(|| format!("Could not open file {}", path.display()))?;
+
+    // If we know the size of the file allocate preemptively its capacity.
+    let file_size = file.metadata().as_ref().map(Metadata::len).unwrap_or(0);
+    let mut bytecode = Vec::with_capacity(file_size as usize);
+    file.read_to_end(&mut bytecode)
+        .with_context(|| format!("Could not read file {}", path.display()))?;
+    let file_name = path
+        .file_name()
+        .expect("Actor file name should be valid")
+        .to_string_lossy()
+        .into_owned();
+    Ok(WasmActor::new(file_name, bytecode, Abi { methods: vec![] }))
+}
+
+/// Gather the target Actor file and its test files.
+/// The rules for reading Actor files and it's matching tests are:
+/// - All .wasm files that are at the root of the kythera input dir are actors.
+/// - All .t.wasm files that are at the root of the kythera wasm dir are test actors.
+/// - All .wasm files that are in .t dirs are test actors.
+pub fn search_files<P: AsRef<Path>>(path: P) -> anyhow::Result<Vec<Test>> {
+    // Search the root dir and find all the .wasm files there which may be target actors
+    // or its matching test dirs and files.
+    // Split into two lists, the first being the target Actors and the second
+    // their matching test files and dirs.
+    let (target_actor_paths, mut test_artifacts_paths): (Vec<String>, Vec<String>) =
+        fs::read_dir(path)
+            .context("Could not read the input path")?
+            .into_iter()
+            .filter_map(Result::ok)
+            // Path::ends_with is diferent from String::ends_with,
+            // Path::ends_with operates on the child, in this case
+            // we don't know the name of the file so we can't operate on the child.
+            .filter_map(|e| e.path().into_os_string().into_string().ok())
+            .filter(|path| path.ends_with(".wasm") || path.ends_with(".t"))
+            .partition(|path| path.ends_with(".wasm") && !path.ends_with(".t.wasm"));
+
+    let mut tests = vec![];
+    for target_actor_path in target_actor_paths {
+        let Ok(main_actor) = read_actor(&target_actor_path) else {
```

Comment:
the `let-else` syntax still looks experimental, am I right?
should we add the dedicated feature gate then?

Reply:
my bad 👇

> the feature `let_else` has been stable since 1.65.0 and no longer requires an attribute to enable

marking as resolved 🙏